pkg/runctl/controller.go

```diff
@@ -868,3 +872,56 @@ func (c *Controller) addToWorkqueueFromAssociated(obj interface{}) {
 		)
 	}
 }
+
+func (c *Controller) finalizePipelineRun(ctx context.Context, pipelineRun k8s.PipelineRun) error {
+
+	if !pipelineRun.HasFinalizer() {
+		return nil
+	}
+
+	err := pipelineRun.DeleteFinalizerAndCommitIfExists(ctx)
+	if err != nil {
+		return err
+	}
+
+	c.logFinalState(ctx, pipelineRun)
+
+	return nil
+}
+
+func (c *Controller) logFinalState(ctx context.Context, pipelineRun k8s.PipelineRun) error {
+	status := pipelineRun.GetStatus()
+	spec := pipelineRun.GetSpec()
+	runID := "unknown"
+	if spec != nil && spec.Logging != nil && spec.Logging.Elasticsearch != nil {
+		err := error(nil)
+		runID, err = toJSONString(&spec.Logging.Elasticsearch.RunID)
+		if err != nil {
+			return errors.WithMessage(err,
+				"could not serialize spec.logging.elasticsearch.runid to JSON",
+			)
+		}
+	}
+
+	logger := klog.FromContext(ctx)
+
+	logger.V(3).Info(
+		"Completed processing of pipeline run",
+		"intent", spec.Intent,
+		"loggingRunId", runID,
+		"startedAt", status.StartedAt,
+		"finishedAt", status.FinishedAt,
+		"result", status.Result,
+		"statusMessage", status.Message,
+		"stateHistory", status.StateHistory,
+	)
+	return nil
+}
+
+func toJSONString(value interface{}) (string, error) {
+	bytes, err := json.Marshal(value)
+	if err != nil {
+		return "", errors.Wrapf(err, "error while serializing to JSON: %v", err)
+	}
+	return string(bytes), nil
+}
```

Comment:
Move this function into another source file in the same package, e.g. `utils.go`.

Reply:
Actually, I took this function from [pkg/runctl/runmgr/run_manager.go#L833](https://github.com/SAP/stewardci-core/blob/PR_final_logging/pkg/runctl/runmgr/run_manager.go#L833).
Maybe we could move both functions to [pkg/utils/stringUtils.go](https://github.com/SAP/stewardci-core/blob/PR_final_logging/pkg/utils/stringUtils.go) to avoid code duplication?

Reply:
Yes, it can be moved to `pkg/utils`, but not into `stringUtils.go` (this is for utils working on strings). Maybe `json.go`.